File: app/src/main/java/com/blogspot/e_kanivets/moneytracker/adapter/RecordAdapter.java
package com.blogspot.e_kanivets.moneytracker.adapter;

import android.annotation.SuppressLint;
import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.BaseAdapter;
import android.widget.TextView;

import com.blogspot.e_kanivets.moneytracker.MtApp;
import com.blogspot.e_kanivets.moneytracker.R;
import com.blogspot.e_kanivets.moneytracker.controller.FormatController;
import com.blogspot.e_kanivets.moneytracker.entity.data.Record;

import java.util.List;

import javax.inject.Inject;

import butterknife.BindView;
import butterknife.ButterKnife;

/**
 * Custom adapter class for {@link Record} entity.
 * Created on 01/09/14.
 *
 * @author Evgenii Kanivets
 */
public class RecordAdapter extends BaseAdapter {
    @Inject
    FormatController formatController;

    private Context context;
    private List<Record> records;

    private int whiteRed;
    private int whiteGreen;
    private int red;
    private int green;

    @SuppressWarnings("deprecation")
    public RecordAdapter(Context context, List<Record> records) {
        MtApp.get().getAppComponent().inject(RecordAdapter.this);

        this.context = context;
        this.records = records;

        whiteRed = context.getResources().getColor(R.color.white_red);
        whiteGreen = context.getResources().getColor(R.color.white_green);
        red = context.getResources().getColor(R.color.red);
        green = context.getResources().getColor(R.color.green);
    }

    @Override
    public int getCount() {
        return records.size();
    }

    @Override
    public Record getItem(int position) {
        return records.get(position);
    }

    @Override
    public long getItemId(int position) {
        return position;
    }

    @SuppressLint("SetTextI18n")
    @Override
    public View getView(final int position, View convertView, ViewGroup parent) {
        ViewHolder viewHolder;

        if (convertView == null) {
            LayoutInflater layoutInflater = LayoutInflater.from(context);

            convertView = layoutInflater.inflate(R.layout.view_record, parent, false);
            viewHolder = new ViewHolder(convertView);

            convertView.setTag(viewHolder);
        } else viewHolder = (ViewHolder) convertView.getTag();

        Record record = getItem(position);
        viewHolder.container.setBackgroundColor(record.isIncome() ? whiteGreen : whiteRed);
        viewHolder.tvPrice.setTextColor(record.isIncome() ? green : red);

        viewHolder.tvDateAndTime.setText(formatController.formatDateAndTime(record.getTime()));
        viewHolder.tvPrice.setText(formatController.formatSignedAmount(
                (record.isIncome() ? 1 : -1) * record.getFullPrice()));
        viewHolder.tvTitle.setText(record.getTitle());
        if (record.getCategory() != null)
            viewHolder.tvCategory.setText(record.getCategory().getName());
        viewHolder.tvCurrency.setText(record.getCurrency());

        return convertView;
    }

    public static class ViewHolder {
        @BindView(R.id.container)
        View container;
        @BindView(R.id.tv_date_and_time)
        TextView tvDateAndTime;
        @BindView(R.id.tv_price)
        TextView tvPrice;
        @BindView(R.id.tv_title)
        TextView tvTitle;
        @BindView(R.id.tv_category)
        TextView tvCategory;
        @BindView(R.id.tvCurrency)
        TextView tvCurrency;

        public ViewHolder(View view) {
            ButterKnife.bind(this, view);
        }
    }
}

File: app/src/main/java/com/blogspot/e_kanivets/moneytracker/adapter/RecordAdapter.kt
package com.blogspot.e_kanivets.moneytracker.adapter

import android.content.Context
import android.support.v4.content.ContextCompat
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import com.blogspot.e_kanivets.moneytracker.MtApp
import com.blogspot.e_kanivets.moneytracker.R
import com.blogspot.e_kanivets.moneytracker.controller.FormatController
import com.blogspot.e_kanivets.moneytracker.entity.RecordItem
import kotlinx.android.synthetic.main.view_header_date.view.*
import kotlinx.android.synthetic.main.view_record.view.*
import javax.inject.Inject

class RecordAdapter : RecyclerView.Adapter<RecyclerView.ViewHolder> {

    @Inject
    lateinit var formatController: FormatController

    private var itemClickListener: ((Int) -> Unit)? = null

    private var whiteRed: Int
    private var whiteGreen: Int
    private var red: Int
    private var green: Int

    private var items: List<RecordItem>
    private var context: Context

    private var isSummaryViewNeeded: Boolean = false
    private lateinit var summaryViewHolder: RecyclerView.ViewHolder

    constructor(context: Context, items: List<RecordItem>, isSummaryViewNeeded: Boolean) {
        this.context = context
        this.items = items

        MtApp.get().appComponent.inject(this)

        whiteRed = ContextCompat.getColor(context, R.color.white_red)
        whiteGreen = ContextCompat.getColor(context, R.color.white_green)
        red = ContextCompat.getColor(context, R.color.red)
        green = ContextCompat.getColor(context, R.color.green)

        this.isSummaryViewNeeded = isSummaryViewNeeded
    }

    fun addSummaryView(summaryView: View) {
        this.summaryViewHolder = summaryView.tag as RecyclerView.ViewHolder
    }

    fun setOnItemClickListener(itemClickListener: ((Int) -> Unit)) {
        this.itemClickListener = itemClickListener
    }

    override fun getItemCount() = items.size + if (isSummaryViewNeeded) 1 else 0

    override fun getItemViewType(position: Int): Int = if (position == 0 && isSummaryViewNeeded) {
        TYPE_SUMMARY
    } else if (items[position - if (isSummaryViewNeeded) 1 else 0] is RecordItem.Header) {
        TYPE_HEADER
    } else {
        TYPE_RECORD
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder =
            when (viewType) {
                TYPE_RECORD -> RecordViewHolder(LayoutInflater.from(context).inflate(R.layout.view_record, parent, false), itemClickListener)
                TYPE_HEADER -> HeaderViewHolder(LayoutInflater.from(context).inflate(R.layout.view_header_date, parent, false))
                else -> summaryViewHolder
            }

    override fun onBindViewHolder(viewHolder: RecyclerView.ViewHolder, position: Int) {
        if (position == 0 && isSummaryViewNeeded) {
            //view holder already bound to view
            return
        }

        if (viewHolder is RecordViewHolder) {
            val record = items[position - if (isSummaryViewNeeded) 1 else 0] as RecordItem.Record
            viewHolder.tvPrice.setTextColor(if (record.isIncome) green else red)

            val price = (if (record.isIncome) record.fullPrice else getNegative(record.fullPrice))
            viewHolder.tvPrice.text = formatController.formatSignedAmount(price)
            viewHolder.tvTitle.text = record.title
            viewHolder.tvCategory.text = record.categoryName
            viewHolder.tvCurrency.text = record.currency
        } else {
            val headerViewHolder = viewHolder as HeaderViewHolder
            val header = items[position - if (isSummaryViewNeeded) 1 else 0] as RecordItem.Header
            headerViewHolder.tvDate.text = header.date
        }
    }

    private fun getNegative(number: Double): Double {
        return -1 * number
    }

    fun setRecords(itemsList: List<RecordItem>) {
        items = itemsList
        notifyDataSetChanged()
    }

    class RecordViewHolder : RecyclerView.ViewHolder {

        var tvPrice: TextView
        var tvTitle: TextView
        var tvCategory: TextView
        var tvCurrency: TextView

        constructor(view: View, itemClickListener: ((Int) -> Unit)?) : super(view) {
            tvPrice = view.tvPrice
            tvTitle = view.tvTitle
            tvCategory = view.tvCategory
            tvCurrency = view.tvCurrency

            view.setOnClickListener {
                itemClickListener?.invoke(adapterPosition)
            }
        }
    }

    class HeaderViewHolder(view: View) : RecyclerView.ViewHolder(view) {
        val tvDate: TextView = view.tvDate
    }

    companion object {
        private const val TYPE_SUMMARY = 0
        private const val TYPE_HEADER = 1
        private const val TYPE_RECORD = 2
    }
}
